From 472b7e6ad071bcdb42d3b4c6da5f7d60b5b18a5e Mon Sep 17 00:00:00 2001
From: Yu Qiaowei <cerf.yu@rock-chips.com>
Date: Wed, 21 Jan 2026 17:46:27 +0800
Subject: [PATCH 28/87] video: rockchip: rga3: mpi_commit: add protection when
 scale-up

Rockit-ko allows user-space to configure interpolation algorithms for
scaling. However, for upscaling, the default algorithm must be enforced,
ignoring user-space configuration.

Signed-off-by: Yu Qiaowei <cerf.yu@rock-chips.com>
Change-Id: I3b5676a06d25ed8c47aa931c4268e95eece492ea
(cherry picked from commit ad7b96ca90faa99146fd5fe904ae63d54b8bdaaa)
---
 drivers/video/rockchip/rga3/rga_drv.c | 28 +++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/drivers/video/rockchip/rga3/rga_drv.c b/drivers/video/rockchip/rga3/rga_drv.c
index 6af4e607aef76..7272111769e7d 100644
--- a/drivers/video/rockchip/rga3/rga_drv.c
+++ b/drivers/video/rockchip/rga3/rga_drv.c
@@ -104,6 +104,32 @@ static void rga_mpi_set_channel_info(uint32_t flags_mask, uint32_t flags,
 	}
 }
 
+static void rga_mpi_scale_protect(struct rga_req *msg)
+{
+	uint32_t rotate_mode;
+	uint32_t sw, sh;
+	uint32_t dw, dh;
+
+	rotate_mode = msg->rotate_mode & 0x3;
+
+	sw = msg->src.act_w;
+	sh = msg->src.act_h;
+
+	if ((rotate_mode == 1) | (rotate_mode == 3)) {
+		dw = msg->dst.act_h;
+		dh = msg->dst.act_w;
+	} else {
+		dw = msg->dst.act_w;
+		dh = msg->dst.act_h;
+	}
+
+	/* when scale-up, using default interpolation algorithm. */
+	if (sw < dw)
+		msg->interp.horiz = RGA_INTERP_DEFAULT;
+	if (sh < dh)
+		msg->interp.verti = RGA_INTERP_DEFAULT;
+}
+
 int rga_mpi_commit(struct rga_mpi_job_t *mpi_job)
 {
 	int ret = 0;
@@ -214,6 +240,8 @@ int rga_mpi_commit(struct rga_mpi_job_t *mpi_job)
 		}
 	}
 
+	rga_mpi_scale_protect(&mpi_cmd);
+
 	mpi_cmd.handle_flag = 1;
 	mpi_cmd.mmu_info.mmu_en = 0;
 	mpi_cmd.mmu_info.mmu_flag = 0;
